### PR TITLE
Updated GencodeGtfCodec to be more permissive.

### LINKED ISSUE #7054: IndexFeatureFile Error to Run Funcotator with Mouse Ensembl GTF
## Description
A user wants to run Funcotator with a mouse sample but has an issue running IndexFeatureFile. The GencodeGTF parser is specific to human data. Funcotator could be useful for users with non-human data if there is a workaround for these errors.

### GATK Information
GATK 4.1.9.0
gatk IndexFeatureFile -I gencode.vM25.annotation.gtf
This request was created from a contribution made by T. Li on January 25, 2021 04:37 UTC.

Link: [https://gatk.broadinstitute.org/hc/en-us/community/posts/360076815852-Error-Running-IndexFeatureFile-on-Ensembl-Mouse-GTF-file-](https://gatk.broadinstitute.org/hc/en-us/community/posts/360076815852-Error-Running-IndexFeatureFile-on-Ensembl-Mouse-GTF-file-)

#### Error Log

```
Using GATK jar /gatk/gatk-package-4.1.9.0-SNAPSHOT-local.jar  
  
Running:  
  
java -Dsamjdk.use\_async\_io\_read\_samtools=false -Dsamjdk.use\_async\_io\_write\_samtools=true -Dsamjdk.use\_async\_io\_write\_tribble=false -Dsamjdk.compression\_level=2 -jar /gatk/gatk-package-4.1.9.0-SNAPSHOT-local.jar IndexFeatureFile -I gencode/mm10/gencode.vM25.annotation.gtf  
  
04:33:13.081 INFO NativeLibraryLoader - Loading libgkl\_compression.so from jar:file:/gatk/gatk-package-4.1.9.0-SNAPSHOT-local.jar!/com/intel/gkl/native/libgkl\_compression.so  
  
Jan 25, 2021 4:33:13 AM shaded.cloud\_nio.com.google.auth.oauth2.ComputeEngineCredentials runningOnComputeEngine  
  
INFO: Failed to detect whether we are running on Google Compute Engine.  
  
04:33:13.195 INFO IndexFeatureFile - ------------------------------------------------------------  
  
04:33:13.195 INFO IndexFeatureFile - The Genome Analysis Toolkit (GATK) v4.1.9.0-SNAPSHOT  
  
04:33:13.195 INFO IndexFeatureFile - For support and documentation go to [https://software.broadinstitute.org/gatk/](https://software.broadinstitute.org/gatk/)  
  
04:33:13.195 INFO IndexFeatureFile - Executing as root@b4c480938d0d on Linux v5.4.0-1029-aws amd64  
  
04:33:13.195 INFO IndexFeatureFile - Java runtime: OpenJDK 64-Bit Server VM v1.8.0\_242-8u242-b08-0ubuntu3~18.04-b08  
  
04:33:13.195 INFO IndexFeatureFile - Start Date/Time: January 25, 2021 4:33:13 AM GMT  
  
04:33:13.195 INFO IndexFeatureFile - ------------------------------------------------------------  
  
04:33:13.195 INFO IndexFeatureFile - ------------------------------------------------------------  
  
04:33:13.196 INFO IndexFeatureFile - HTSJDK Version: 2.23.0  
  
04:33:13.196 INFO IndexFeatureFile - Picard Version: 2.23.3  
  
04:33:13.196 INFO IndexFeatureFile - HTSJDK Defaults.COMPRESSION\_LEVEL : 2  
  
04:33:13.196 INFO IndexFeatureFile - HTSJDK Defaults.USE\_ASYNC\_IO\_READ\_FOR\_SAMTOOLS : false  
  
04:33:13.196 INFO IndexFeatureFile - HTSJDK Defaults.USE\_ASYNC\_IO\_WRITE\_FOR\_SAMTOOLS : true  
  
04:33:13.196 INFO IndexFeatureFile - HTSJDK Defaults.USE\_ASYNC\_IO\_WRITE\_FOR\_TRIBBLE : false  
  
04:33:13.196 INFO IndexFeatureFile - Deflater: IntelDeflater  
  
04:33:13.196 INFO IndexFeatureFile - Inflater: IntelInflater  
  
04:33:13.196 INFO IndexFeatureFile - GCS max retries/reopens: 20  
  
04:33:13.196 INFO IndexFeatureFile - Requester pays: disabled  
  
04:33:13.196 INFO IndexFeatureFile - Initializing engine  
  
04:33:13.196 INFO IndexFeatureFile - Done initializing engine  
  
04:33:13.396 INFO FeatureManager - Using codec EnsemblGtfCodec to read file file:///gatk/funcotator-scripts/gencode/mm10/gencode.vM25.annotation.gtf  
  
04:33:13.400 INFO ProgressMeter - Starting traversal  
  
04:33:13.400 INFO ProgressMeter - Current Locus Elapsed Minutes Records Processed Records/Minute  
  
04:33:21.040 INFO IndexFeatureFile - Shutting down engine  
  
\[January 25, 2021 4:33:21 AM GMT\] org.broadinstitute.hellbender.tools.IndexFeatureFile done. Elapsed time: 0.13 minutes.  
  
Runtime.totalMemory()=1835532288  
  
java.lang.IllegalArgumentException: Unexpected value: IG\_D\_pseudogene  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.GencodeGtfFeature$GeneTranscriptType.getEnum(GencodeGtfFeature.java:1060)  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.GencodeGtfFeature.<init>(GencodeGtfFeature.java:158)  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.GencodeGtfGeneFeature.<init>(GencodeGtfGeneFeature.java:19)  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.GencodeGtfGeneFeature.create(GencodeGtfGeneFeature.java:23)  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.GencodeGtfFeature$FeatureType$1.create(GencodeGtfFeature.java:760)  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.GencodeGtfFeature.create(GencodeGtfFeature.java:327)  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.AbstractGtfCodec.decode(AbstractGtfCodec.java:138)  
  
at org.broadinstitute.hellbender.utils.codecs.gtf.AbstractGtfCodec.decode(AbstractGtfCodec.java:23)  
  
at htsjdk.tribble.AbstractFeatureCodec.decodeLoc(AbstractFeatureCodec.java:43)  
  
at org.broadinstitute.hellbender.utils.codecs.ProgressReportingDelegatingCodec.decodeLoc(ProgressReportingDelegatingCodec.java:46)  
  
at htsjdk.tribble.index.IndexFactory$FeatureIterator.readNextFeature(IndexFactory.java:689)  
  
at htsjdk.tribble.index.IndexFactory$FeatureIterator.next(IndexFactory.java:650)  
  
at htsjdk.tribble.index.IndexFactory.createIndex(IndexFactory.java:511)  
  
at htsjdk.tribble.index.IndexFactory.createDynamicIndex(IndexFactory.java:446)  
  
at org.broadinstitute.hellbender.tools.IndexFeatureFile.createAppropriateIndexInMemory(IndexFeatureFile.java:118)  
  
at org.broadinstitute.hellbender.tools.IndexFeatureFile.doWork(IndexFeatureFile.java:75)  
  
at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:140)  
  
at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:192)  
  
at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:211)  
  
at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)  
  
at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)  
  
at org.broadinstitute.hellbender.Main.main(Main.java:289)<br><br><i>(created from <a href='https://broadinstitute.zendesk.com/agent/tickets/100645'>Zendesk ticket #100645</a>)<br>gz#100645</i>
```

### PR DESCRIPTION
**Pull request by jonn-smith**
- Now the `GencodeGtfCodec` no longer parses `transcriptType` and `geneType`
into enums.  These fields are now each stored as a `String`.  This allows for arbitrary
values in these fields and will help to future-proof (and species proof) the GTF parser.

- Fixes #7054

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/TranscriptSelectionMode.java [hunk 1]
```diff
@@ -246,8 +246,8 @@ public ComparatorByProteinCodingStatus(){}
         @Override
         public int compare( final GencodeFuncotation a, final GencodeFuncotation b ) {
             // Is it protein coding?
-            final boolean isAProteinCoding = a.getGeneTranscriptType() == GencodeGtfFeature.GeneTranscriptType.PROTEIN_CODING;
-            final boolean isBProteinCoding = b.getGeneTranscriptType() == GencodeGtfFeature.GeneTranscriptType.PROTEIN_CODING;
+            final boolean isAProteinCoding = (a.getGeneTranscriptType() != null) && a.getGeneTranscriptType().equals(GencodeGtfFeature.KnownGeneBiotype.PROTEIN_CODING.toString());
+            final boolean isBProteinCoding = (b.getGeneTranscriptType() != null) && b.getGeneTranscriptType().equals(GencodeGtfFeature.KnownGeneBiotype.PROTEIN_CODING.toString());
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/TranscriptSelectionMode.java:250]: Minor point: you wouldn't need the explicit null checks here if you reversed the order of the comparison. Eg.,  `GencodeGtfFeature.KnownGeneBiotype.PROTEIN_CODING.toString().equals(b.getGeneTranscriptType())`

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java [hunk 2]
```diff
@@ -1079,11 +1079,11 @@ private GencodeFuncotation createExonFuncotation(final VariantContext variant,
 
         // Before we get started, check to see if this is a non-protein-coding feature.
         // If it is, we must handle it differently:
-        if ( transcript.getGeneType() != GencodeGtfFeature.GeneTranscriptType.PROTEIN_CODING) {
-            return createCodingRegionFuncotationForNonProteinCodingFeature(variant, altAllele, reference, transcript, exon);
+        if ( transcript.getGeneType().equals(GencodeGtfFeature.KnownGeneBiotype.PROTEIN_CODING.toString()) ) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java:1082]: Can `transcript.getGeneType()` be null? If so, reverse the order of the comparison (comment also applies to code below)

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java [hunk 3]
```diff
@@ -2736,84 +2736,20 @@ else if ( gtfApprisTags.size() == 1 ) {
...
-                return GencodeFuncotation.VariantClassification.RNA;
-        }
+     * @param type A {@link String} representing a GeneTranscriptType to convert to a {@link org.broadinstitute.hellbender.tools.funcotator.dataSources.gencode.GencodeFuncotation.VariantClassification}.
+     * @return A {@link org.broadinstitute.hellbender.tools.funcotator.dataSources.gencode.GencodeFuncotation.VariantClassification} representing the given GeneTranscriptType {@link String}, or {@code null}.
+     */
+    private static GencodeFuncotation.VariantClassification convertGeneTranscriptTypeToVariantClassification (final String type ) {
+
+        //TODO: This all needs to be fixed so there is a 1:1 mapping of GencodeGtfFeature.KnownGeneBiotype->VariantClassification - Issue #4405
+        if (type.equals(GencodeGtfFeature.KnownGeneBiotype.LINCRNA.toString()) ||
+            type.equals(GencodeGtfFeature.KnownGeneBiotype.MACRO_LNCRNA.toString())) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java:2749]: Same comment -- if `type` can be be null reverse the order of the comparison (ie., `GencodeGtfFeature.KnownGeneBiotype.LINCRNA.toString().equals(type)`)

File: src/test/java/org/broadinstitute/hellbender/utils/codecs/gtf/GencodeGtfCodecUnitTest.java [hunk 4]
```diff
@@ -3560,6 +3563,21 @@ private GencodeGtfGeneFeature createGencodeGtfGene_gencode_v19_and_this_is_a_val
...
+
+        final String templateFile = "";
+
+        return new Object[][] {
+                { "GAJUAEWJIWIEG", "gencode.biotype_template.gtf", testResourceDir },
+                { "2e#H!FXh&F!N4xn$VD*e", "gencode.biotype_template.gtf", testResourceDir },
+                { "G23&#MPqP6gveAOcakY2", "gencode.biotype_template.gtf", testResourceDir },
+                { "eGPvPbMcHYSbMoUDd@TY", "gencode.biotype_template.gtf", testResourceDir },
+                { "9SjyZMBAXY=Gg9eAyk62", "gencode.biotype_template.gtf", testResourceDir },
+                { "F9rt#9p9Ggex4dXs=EYO", "gencode.biotype_template.gtf", testResourceDir },
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/codecs/gtf/GencodeGtfCodecUnitTest.java:3577]: Add a comment explaining why you're using weird strings in this DataProvider

**Review by droazen — APPROVED** [review]
@jonn-smith Back to you with a few minor comments. Go ahead and merge after addressing them.

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/TranscriptSelectionMode.java (hunk 1 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/TranscriptSelectionMode.java:250]: Yup.  Fixed.

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java (hunk 2 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java:1082]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java (hunk 3 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java:2749]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java [hunk 5]
```diff
@@ -1700,7 +1700,7 @@ private GencodeFuncotation createIntronFuncotation(final VariantContext variant,
         gencodeFuncotationBuilder.setReferenceContext(referenceBases.getBaseString(Strand.POSITIVE));
 
         // Set the VariantClassification:
-        if ( transcript.getGeneType() == GencodeGtfFeature.GeneTranscriptType.PROTEIN_CODING ) {
+        if ( transcript.getGeneType().equals(GencodeGtfFeature.KnownGeneBiotype.PROTEIN_CODING.toString()) ) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java:1703]: @jonn-smith Looks like this one was missed

Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/gencode/GencodeFuncotationFactory.java]: Oops.  Fixed.

**Pull request merged**